File: 0076-minimum-window-substring/solution.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not t or not s:
            return "" # Fix: Return empty string, not False
            
        count, window = {}, Counter(t)
        res, resLen = [-1, -1], float("inf")
        l = 0
        have, need = 0, len(window)
        
        for r in range(len(s)):
            c = s[r]
            count[c] = 1 + count.get(c, 0)

            if c in window and count[c] == window[c]:
                have += 1
            
            while have == need:
                # update our result
                if (r - l + 1) < resLen:
                    res = [l, r]
                    resLen = (r - l + 1)
                
                # pop from the left of our window
                # TODO 3: Decrement count[s[l]] (not remove)
                count[s[l]] -= 1 
                
                # TODO 4: If s[l] in window and count[s[l]] < window[s[l]], decrement 'have'
                if s[l] in window and count[s[l]] < window[s[l]]:
                    have -= 1
                l += 1
                
        l, r = res
        return s[l : r+1] if resLen != float("inf") else ""

File: 0076-minimum-window-substring/solution.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (t.size() == 0){
            return "";
        }
        unordered_map<char, int> needmap;
        for (char cur:t){
            needmap[cur]++;
        }
        int need = needmap.size();
        int have = 0;
        unordered_map<char, int> havemap;
        int reslen = INT_MAX;
        int minLen = 0;
        int l = 0;
        for(int r = 0; r < s.size(); ++r){
            havemap[s[r]]++;
            if (havemap[s[r]] == needmap[s[r]]){
                have++;
            }
            while(have == need){
                if ((r - l + 1) < reslen){
                    reslen = (r - l + 1);
                    minLen = l;
                }
                havemap[s[l]] -= 1;
                if (havemap[s[l]] < needmap[s[l]]){
                    have--;
                }
                l++;
            }

        }
        if (reslen == INT_MAX){
            return "";
        }
        else{
            return s.substr(minLen, reslen);
        }

    }
};
